### DISCUSSION TITLE
How to get the CEK to decrypt the ID Token?

### DISCUSSION THREAD

**Discussion by lrodriguez14** [keycloak/keycloak#9990] [Q&A] [answered]
Hello, I was able to encrypt my id token with Keycloak, but now I am trying to understand how to decrypt it?

Following the documentation of [Signed and Encrypted ID Token Support](https://www.keycloak.org/docs/latest/server_admin/index.html#con-advanced-settings_server_administration_guide) it says that **Keycloak and a client must negotiate which CEK is used and how it is delivered** but how can I negotiate with it? And where can I get the CEK?

**Comment by tnorimat** [comment, marked as answer]
@lrodriguez14 Hello, current Keycloak only supports [Key Encryption](https://datatracker.ietf.org/doc/html/rfc7518#section-4) for negotiating which CEK is used. How it works is as follows.

- Keycloak generates a CEK.
- Keycloak encrypts an ID Token by the CEK (algorithm "enc" = A128GCM, A192GCM, A256GCM, A128CBC-HS256, A192CBC-HS384, A256CBC-HS512 defined in [JWA](https://datatracker.ietf.org/doc/html/rfc7518#section-5)).
- Keycloak encrypts the CEK by using client app's public key (algorithm "alg" = RSA1_5, RSA-OAEP or RSA-OAEP-256 defined in [JWA](https://datatracker.ietf.org/doc/html/rfc7518#section-4)).
- Keycloak constructs an ID Token as JWE.
- Client App gets an encrypted CEK from a JWE encoded ID token and decrypt CED by using its private key (algorithm "alg" = RSA1_5, RSA-OAEP or RSA-OAEP-256 defined in [JWA](https://datatracker.ietf.org/doc/html/rfc7518#section-4)).
- Client App decrypts the JWE encoded ID token by using the CEK (algorithm "enc" = A128GCM, A192GCM, A256GCM, A128CBC-HS256, A192CBC-HS384, A256CBC-HS512 defined in [JWA](https://datatracker.ietf.org/doc/html/rfc7518#section-5)).

How a client app tells its public key used for encrypting CEK to Keycloak is described in [Server Administration Guide](https://www.keycloak.org/docs/latest/server_admin/index.html#con-advanced-settings_server_administration_guide).
Namely, using JWKS_URI. JWKS also might be work. Please refer to Keys tab of a client on Admin Console.
